### ISSUE TITLE
Many ctags warnings due to kind name redefinition

### ISSUE THREAD

**Issue by ohm314** [oracle/opengrok#3764]
**Describe the bug**

While looking through indexer logs I keep stumbling over warning lines like this:

```
2021-11-18 03:20:35.063+0000 WARNING t27 Ctags.lambda$run$0: Error from ctags: ctags: Warning: Don't reuse the kind letter `c' in a language Haskell (old: "constructor", new: "classes")
```

Those are just warnings, so could be ignored, but I was wondering whether they could be fixed before they lead to errors down the line (similar to #1137). For above example, the Haskell definition is overridden by opengrok here: https://github.com/oracle/opengrok/blob/00e4c74bf30151dafe908105e79986013b8671c1/opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/Ctags.java#L357
More precisely, line 362 is defining a new regex to tag a constructor (or class). The problem is with the kind definition, the code uses the extended form, but changed the long name from the one defined in ctags:

```
/usr/local/bin/ctags --list-kinds-full | grep Haskell
Haskell              c      constructor           yes     no      0      NONE   type constructors
Haskell              f      function              yes     no      0      NONE   functions
Haskell              m      module                yes     no      0      NONE   modules
Haskell              t      type                  yes     no      0      NONE   types
```
Looking at ctags help I see that the full kind name is optional and its use deprecated, but I am not sure if actually the redefined name is necessary for Opengrok.

(From ctags documentation)
> A kind specifier (<kind-spec>) for tags matching regexp may follow `<name_pattern>`, which will determine what kind of tag is reported in the kind extension field (see tags(5)).
>
> `<kind-spec>` has two forms: one-letter form and full form.
>
> The one-letter form in the form of `<letter>`. It just refers a kind `<letter>` defined with --kinddef-<LANG>. This form is recommended in Universal Ctags.
>
> The full form of `<kind-spec>` is in the form of `<letter>,<name>,<description>`. Either the kind `<name>` and/or the `<description>` can be omitted. See the description of `--kinddef-<LANG>=<letter>,<name>,<description>` option about the elements.
>
> The full form is supported only for keeping the compatibility with Exuberant Ctags which does not have `--kinddef-<LANG> `option. Supporting the form will be removed from Universal Ctags in the future.

**Versions:**
Opengrok: 1.7.18
ctags: 5.9.0

**Possible Solution**

If the long names is not necessary, they should be just removed from all `--regex` definitions:

So for Haskell this would be:

```java
        command.add("--regex-haskell=/^[[:space:]]*class[[:space:]]+([a-zA-Z0-9_]+)/\\1/c/");
        command.add("--regex-haskell=/^[[:space:]]*data[[:space:]]+([a-zA-Z0-9_]+)/\\1/t/");
        command.add("--regex-haskell=/^[[:space:]]*newtype[[:space:]]+([a-zA-Z0-9_]+)/\\1/t/");
        command.add("--regex-haskell=/^[[:space:]]*type[[:space:]]+([a-zA-Z0-9_]+)/\\1/t/");
        command.add("--regex-haskell=/^([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\1/f/");
        command.add("--regex-haskell=/[[:space:]]+([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\1/f/");
        command.add("--regex-haskell=/^(let|where)[[:space:]]+([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\2/f/");
        command.add("--regex-haskell=/[[:space:]]+(let|where)[[:space:]]+([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\2/f/");
```
If the intention is also to rename the tag because you need it down the line (sorry that part I don't understand 😄 ), then this should be done via a new kind definition, maybe like this:

```java
        command.add("--kinddef-Haskell=C,classes,Classes");
        command.add("--kinddef-Haskell=T,types,Types");
        command.add("--kinddef-Haskell=F,functions,Functions");
        command.add("--regex-haskell=/^[[:space:]]*class[[:space:]]+([a-zA-Z0-9_]+)/\\1/C/");
        command.add("--regex-haskell=/^[[:space:]]*data[[:space:]]+([a-zA-Z0-9_]+)/\\1/T/");
        command.add("--regex-haskell=/^[[:space:]]*newtype[[:space:]]+([a-zA-Z0-9_]+)/\\1/T/");
        command.add("--regex-haskell=/^[[:space:]]*type[[:space:]]+([a-zA-Z0-9_]+)/\\1/T/");
        command.add("--regex-haskell=/^([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\1/F/");
        command.add("--regex-haskell=/[[:space:]]+([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\1/F/");
        command.add("--regex-haskell=/^(let|where)[[:space:]]+([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\2/F/");
        command.add("--regex-haskell=/[[:space:]]+(let|where)[[:space:]]+([a-zA-Z0-9_]+).*[[:space:]]+={1}[[:space:]]+/\\2/F/");
```

This command line is getting extremely long, so maybe all these options should simply be provided in [option file](https://docs.ctags.io/en/latest/option-file.html).

**Comment by ahornace** [comment]
Nice investigation, good job :)

I don't believe it's particularly used somewhere else. I've tested the first approach and it seems to work well. Similar fix will be needed for kotlin and maybe the others? I don't know much about ctags, maybe @vladak will know if there is something to be wary about.

I also encountered a problem with the terraform regex:
```
WARNING: Error from ctags: ctags: Warning: regcomp [[:<:]]resource[[:space:]]*"([[:alpha:]][-_[:alpha:]]*)"[[:space:]]*"([[:alpha:]][-_[:alpha:]]*)"[[:space:]]*\{: Invalid character class name
```

It seems that `[[:<:]]` is not recognized even though it's POSIX. Maybe we could replace it with `^[[:space:]]*` as used in the other regexes.

**Comment by ohm314** [comment]
Yes, absolutely, the patch will need to applied to all language regex options, not only Haskell, and indeed I did come across the terraform regex error but had ignored it for now 😄 .
If you'd like, I can submit a PR with the changes.

**Comment by ahornace** [comment]
> If you'd like, I can submit a PR with the changes.

Sure, that would be awesome! Thanks.

**Comment by ahornace** [comment]
Closed by #3770
